### ISSUE TITLE
Refactor compiler flags

### ISSUE THREAD

**Issue by AnHeuermann** [OpenModelica/OpenModelica#10951]
### Description

There are way to many [OpenModelica Compiler flags](https://openmodelica.org/doc/OpenModelicaUsersGuide/latest/omchelptext.html). The list is huge, it is unordered and mostly unstructured.
Name of flags are not following any system (.e.g. `-d=dump,backenddaeinfo,dumpindxdae,optdaedump,debugDAEmode` are all flags to print some information to the shell) and a lot of flags are simply outdated.
And even deprecated flags are still listed in the documentation.

### Expected Behavior

@kabdelhak and myself propose to change the compiler flags to be sorted into clear categories within a hierarchy.
Duplicate or outdated flags should be deprecated and deprecated flags should be removed.

We need a system to handle deprecation to exclude them from the documentation, tell users what the new flag to use is.
So something like `handleDeprecatedFlags`, but in a more generic implementation.

Not all flags are equally important for every user. We could sort flags for the intended user: End user (uses OMEdit), developer (want super detailed information about some module ), bootstrapping (the compile needs to behave slightly different for building OM).

**Proposed structure**

  - **Logs**
    Everything that provides information to shell / file / etc.
    Is not allowed to have any side-effects.
    - `--log=`
    - Examples:
      -  `-d=backenddaeinfo` --> `--log=backend`
      - `-d=dumprepl,dumpFPrepl,dumpParamrepl,dumpPPrepl,dumpEArepl,debugAlias` --> `--log=alias`
      - `-d=symjacdump,symjacdumpverbose,symjacdumpeqn,symjacwarnings` --> `--log.jacobian.verbose`
      - `-d=dumpAlgebraicLoopsJacobian` --> `--log=jacobian.nonlinear`
      - `-d=dumpJL` --> `--log=absyn.julia`
 - **Warnings**
   Changes the way compiler warnings are displayed or makes errors to warnings (and vice versa)
   - Examples:
     - `-d=failtrace` --> `--warnings=failtrace`
     - `-d=warnNoNominal` --> `--warnings=nominal`
 - **Modelica Standard**
    Select Modelica Standard to use, allow non-standard functionalities
   - Examples:
     - `--std=1.x` --> `--std.version=1.x`
     - `--allowNonStandardModelica` --> `--std.allow=nonStandard`
     - `-d=ignoreCycles` --> `--std.allow=cyclicParameters`
 - **Configurations**
   - `-d=-vectorize,nfScalarize,nfExpandOperations,nfExpandFuncArgs`  -> `--config=scalarize`
   - `-d=nfAPI,nfAPINoise` --> `--config=newAPI`
   - `-d=-disableColoring` --> `--config=jacobian.coloring`
   - `--generateSymbolicJacobian` --> `--config=jacobian=symbolic`
   - `-d=skipInputOutputSyntacticSugar` --> `--config=bootstrapping.skipInputOutput`

### Version and OS

 - OpenModelica Version: v1.22.0-dev-237

**Comment by AnHeuermann** [comment]
So far we picked the log flags from the debug flags and tried to order them:

<details>
<summary>List of all log flags </summary>

## `--log`
Concatenate the levels of the following list to get the new log flag for the old flag.
Some parent levels can be omitted when in `[` `]`, like `[frontend]`.

E.g. `-d=gcProfiling` --> `--log=gc.profiling` or

`-d=cgraphGraphVizFile` --> `--log=frontend.connect.graph.file`
/ `--log=connect.graph.file`.

  - gc
    - profiling: `-d=gcProfiling`

  - statistics
    - modules: `-d=execstat,dumpBackendClocks`
    - template:  `-d=tplPerfTimes`

  - \[frontend\]:  `-d=dumpFlatModel`
    - absyn:  `-d=dump`
      - graphviz: `-d=graphviz`
      - julia: `-d=dumpJL`
    - connect: `-d=dumpSetBasedGraphs`
      - graph
        - file: `-d=cgraphGraphVizFile`
        - show: `-d=cgraphGraphVizShow`
    - parameters
      - structural: `-d=printStructuralParameters`
    - simplify: `-d=checkSimplify,dumpSimplify`
    - units: `-d=dumpUnits`
      - equations: `-d=dumpEqInUC`
      - tree: `-d=dumpEqInUCStruct`

  - \[backend\]: `-d=backenddaeinfo`
    - stateselection: `-d=stateselection`
    - discretes: `-d=discreteinfo`
    - \[modules\]: `-d=dumpindxdae,optdaedump`
      - daemode: `-d=debugDAEmode`
      - lowering: `-d=dumpdaelow`
      - bindings: `-d=dumpBindings`
      - partitioning: `-d=dumpSynchronous`
      - events: `-d=dumpEvents,relidx`
        - when: `-d=dumpEncapsulateConditions`
      - alias: `-d=dumprepl,dumpFPrepl,dumpParamrepl,dumpPPrepl,dumpEArepl,debugAlias`
      - constants: `-d=dumpConstrepl`
      - conflicts: `-d=aliasConflicts`
      - evaluate
        - parameter: `-d=evaluateParamterDump`
      - commonsubexpressions|cse: `-d=dumpCSE`
        - v\[erbose\]: `-d=dumpCSE_verbose`
      - jacobian:
        - v\[erbose\]: `-d=symjacdump,symjacdumpverbose,symjacdumpeqn,symjacwarnings`
        - sparsity: `-d=dumpSparsePattern,dumpSparsePatternVerbose`
        - nonlinear: `-d=dumpAlgebraicLoopsJacobian`
        - linear
        - ode
        - dae
        - splitconst: `-d=symJacConstantSplit`
      - differentiation: `-d=debugDifferentiation`
        - v\[erbose\]: `-d=debugDifferentiationVerbose`
      - causalization
        - v\[erbose\]
        - matching
        - sorting
        - indexreduction: `-d=dummyselect,bltdump`
        - graphml: `-d=graphml`
        - outputOnly: `-d=evalOutputOnly`
        - blt
          - html: `-d=bltmatrixdump`
      - algebraicloops: `-d=dumpLoops`
      - assc: `-d=dumpASSC`
      - tearing: `-d=tearingdump,dumpRecursiveTearing`
        - v\[erbose\]: `-d=tearingdumpVerbose, dumpLoopsVerbose,totaltearingdumpV,totaltearingdump`
        - iterationvariables
      - simplify:
        - loops: `-d=dumpSimplifyLoops`
      - solve: `-d=dumpSlice`
      - inline: `-d=dumpBackendInline,dumpBackendInlineVerbose`
      - initialization: `-d=dumpInitialSystem`
        - balancing: `-d=initialization`
        - parameter: `-d=paramdlowdump`
      - homotopy: `-d=dumpHomotopy`
    - uncertainties: `-d=uncertainties`
      - v\[erbose\]: `-d=dataReconciliation`
    - parallelization: `-d=hpcomDump`
  - simcode: `-d=dumpSimCode`
    - serialized: `-d=reportSerializedSize`
  - \[script\]:
    - conversionRules: `-d=dumpConversionRules`
  - interactive: `-d=interactivedump`

</details>

**Comment by AnHeuermann** [comment]
In addition to collecting a bunch of log flags we found 83 flags that we would like to deprecate and remove:

<details>
<summary>List of log flags to remove:</summary>


  1. ceval
  1. checkBackendDAE
  1. pThreads
  1. events   # Should not be possible to deactivate
  1. dumpInlineSolver   # Unused
  1. evalFunc
  1. gen
  1. generateCodeCHeat
  1. checkDAECrefType
  1. checkASUB
  1. instance
  1. cache
  1. rml
  1. tail
  1. lookup
  1. experimentalReductions
  1. types
  1. transformsbeforedump
  1. daedumpgraphv
  1. interactive
  1. interactiveCorba
  1. interactiveDump
  1. relidx
  1. dumprepl
  1. dumpFPrep
  1. dumpParamrepl
  1. dumpPPrepl
  1. dumpEArepl
  1. debugAlias
  1. tearingdump
  1. execHash
  1. countOperations
  1. upmod
  1. static
  1. dumpeqinorder
  1. dumpsemilinear
  1. showStartOrigin
  1. graphInst
  1. graphInstRunDep
  1. graphInstGenGraph
  1. showEquationSource
  1. dumpInlineSolver
  1. inlineSolver
  1. inlineFunctions 
  1. dumpSCCGraphML
  1. graphvizDump             # New Backend can't do it 
  1. resolveLoopsDump         # replaced by ASSC
  1. disableWindowsPathCheckWarning   # Should not be possible to deactivate
  1. disableRecordConstructorOutput   # Will be logged if logging is on. Take it or leave it...
  1. implOde              # Not used
  1. evalFuncDump         # Deprecated evalFunc
  1. acceptToManyFileds   # Not used in new Frontend
  1. scodeDep
  1. showInstCache
  1. showDaeGeneration    # Not used in new Frontend
  1. reshufflePost        # replaced by ASSC
  1. showExpandableInfo   # Not used in new Frontend
  1. relocatableFunctions # Is this used by anyone?
  1. disableStartCalc     # Always false (no downside?)
  1. constjac             # replaced by ASSC
  1. dumpSortEqnsAndVars  # Irrelevant information
  1. dumpFunctions        # Only used once in the compiler
  1. dumpdgesv            # Runtime decision
  1. multirate            # Not implemented
  1. dumpExcludedSymJacExps # Maybe add to a verbose log flag of Jacobian module
  1. disableJacsforSCC    # Is this used? And will this save time?
  1. forceNLSanalyticJacobian     # Always do for symbolic Jacobians
  1. metaModelicaRecordAllocWords # Not used
  1. parallelCodegen              # Buy a better PC... or use --numProcs
  1. backendKeepEnv               # Sounds horrible
  1. listAppendWrongOrder         # Should always be on, never deactivate
  1. frontEndUnitCheck            # Use --unitChecking (or something like this)
  1. oldFrontEndUnitCheck         # remove
  1. mergeAlgSections             # In general this is wrong (or creates a giant loop)
  1. backendReduceDAE             # Not even tested
  1. execstatGCollect             # No chance that this is used
  1. printRecordTypes             # Not used / always done for new Frontend

Ask @adrpo 
  1. patternmSkipFilterUnusedBindings
  1. patternmAllInfo
  1. patternmDeadCodeElimination
  1. patternmMoveLastExp
  1. writeToBuffer

Ask @peros
  1. stripPrefix  # Always on?
</details>

**Comment by AnHeuermann** [comment]
> IMHO if we want to clean up flag names, we have a unique opportunity to make these flag regular ones, e.g. --evalOutputOnly or --symJacConstantSplit. What do you think?

I agree. The idea of @kabdelhak and myself was to start with the "log" flags and move them and then categorize other flags.
We could have one category for flags that influence how simulation code is generated. One category for flags that control frontend or backend modules and so on.

**Comment by casella** [comment]
That's one option, but is it really necessary? I won't introduce non-backwards compatible changes unless there is a clear advantage in doing so

**Comment by AnHeuermann** [comment]
> That's one option, but is it really necessary?

No. But it's just so  inconvenient to use the log flags, both for users and developers.

We thought about leaving the old flags for now, but they will print a deprecation warning (and activate the new flag) and in OM 2.0 we can remove all outdated / changed flags for good.

**Comment by casella** [comment]
Sounds good. But let's see the whole finished thing together once, before starting to deprecate stuff 😅
